### PR TITLE
Add line, function name for debug file

### PR DESCRIPTION
**Pull request by danghai**
Hi Nick, I add the line, and function name to tlog-test-timespec.c to provide information for debug. Can you look at it? If it is fine, I will modify the rest of test C file. Because I want to know exact line of test case that makes the code error, I add __LINE__ as a parameter in function op_list_test(....). I am not sure it is a good way or not. If you know a better way to do this, please let me know. I give the image modify the test case (116-119 and 124-127).  I also want to list exact operation function pointer that makes wrong case, but I do not know how to find the good way. In my code it only reports op_list_test function. 
debug

### CODE REVIEW

File: src/tlog-test-timespec.c [hunk 1]
```diff
@@ -48,15 +48,16 @@ op_list_test(const struct op *op_list, size_t op_num,
         passed = (res.tv_sec == exp_res.tv_sec && res.tv_nsec == exp_res.tv_nsec);
         if (passed) {
             fprintf(stderr,
-                    "PASS " TLOG_TIMESPEC_FMT " %s " TLOG_TIMESPEC_FMT
+                    "[PASS] " TLOG_TIMESPEC_FMT " %s " TLOG_TIMESPEC_FMT
```
Comment (spbnick) [inline src/tlog-test-timespec.c]: Let's not add the `[]' decorations just for one test and together with adding another change. I would prefer to not add them in general, as I think when we need to decorate text, then it means we failed producing a readable output otherwise, and it's better to improve/simplify the output.

File: src/tlog-test-timespec.c [hunk 2]
```diff
@@ -48,15 +48,16 @@ op_list_test(const struct op *op_list, size_t op_num,
...
                     " == " TLOG_TIMESPEC_FMT "\n",
                     TLOG_TIMESPEC_ARG(&a), op_list[i].sym,
                     TLOG_TIMESPEC_ARG(&b),
                     TLOG_TIMESPEC_ARG(&exp_res));
         } else {
             fprintf(stderr,
-                    "FAIL " TLOG_TIMESPEC_FMT " %s " TLOG_TIMESPEC_FMT
+                    "[FAIL] %10s:%10d:%15s(): " TLOG_TIMESPEC_FMT " %s " TLOG_TIMESPEC_FMT
                     " = " TLOG_TIMESPEC_FMT " != " TLOG_TIMESPEC_FMT "\n",
+                    __FILE__,  line, __func__,
```
Comment (spbnick) [inline src/tlog-test-timespec.c]: I think we can do without `__func__` here, especially since it's always the same. Then, to be uniform, please pass the file name as an argument too. Also, please don't pad filename and line number separately. Instead join them with a colon in a separate buffer, and pad that buffer to the right in this `fprintf` here. That way we'll save some horizontal space.

File: src/tlog-test-timespec.c [hunk 3]
```diff
@@ -98,200 +99,200 @@ main(void)
 
 #define TS(_sec, _nsec) (struct timespec){_sec, _nsec}
 
-#define TEST(_op_list, _a, _b, _exp_res) \
+#define TEST(_op_list, _a, _b, _exp_res, line) \
```
Comment (spbnick) [inline src/tlog-test-timespec.c]: Don't add the `line` argument here. Simply use `__LINE__` (and `__FILE__`) in the macro body and they will be turned into the file name and line number of the place where the macro was expanded, by the compiler, and you won't need to pass them every time, below.

**Comment by danghai** [comment]
Do you mean that I simply change the same code below? : 
https://github.com/danghai/tlog/blob/tlog_debug/src/tlog-test-timestr.c#L51

It actually produces the same line error for every mistake. That is reason why I put line as a parameter. Or  do I miss something?

**Comment by spbnick** [comment]
No, I meant put the `__FILE__` and `__LINE__` into the `TEST` macro body, where the `test` function is invoked.

**Comment by danghai** [comment]
I understand what you mean. I add the if-condition in `TEST` macro. Is it good now?

**Comment by spbnick** [comment]
Thank you, Hai. You got the part where `__FILE__` and `__LINE__` should be in the macro body, but don't put an extra condition there as well. Just pass `__FILE__' and `__LINE__` to the testing functions as arguments. This way you get correct file name and line number, and get to print them in the same place as the rest of the test output.

**Comment by spbnick** [comment]
Like this:

    passed = test(__FILE__, __LINE__, _timestr, _exp_rc, _exp_ts)

**Comment by danghai** [comment]
[test-suite.log](https://github.com/Scribery/tlog/files/1524480/test-suite.log)

**Comment by spbnick** [comment]
Thank you, Hai, we're getting there!

Alright. First two things are: please don't add or change non-significant whitespace in your patches, unless they're specifically about that (such as reformatting the code), and please keep the lines of code you add or edit under 78 characters long. The former ensures that patches are easier to read, as they don't have irrelevant changes, and the latter makes the code easier to read in general.

Now onto your actual changes. I think since we're updating all the tests to add file and line information, we can as well make them uniform. How about we make all the tests *always* output file name and line number, no matter if test passes or fails, and do it in in the same format. Like this:

    STATUS FILE_NAME:LINE_NUMBER: TEST_NAME_AND_INFO

Since the status string is always the same width (`PASS`/`FAIL`), the file name is always the same within the test so far, and line number doesn't change considerably, we'll get the test name aligned on the same column, or very close, making the output easier to read.

E.g. something like this:

	PASS tlog-test-json-overlay.c:102: empty
	PASS tlog-test-json-overlay.c:107: atomic_lower
	PASS tlog-test-json-overlay.c:112: atomic_upper
	PASS tlog-test-json-overlay.c:117: same_atomic_both
	PASS tlog-test-json-overlay.c:122: different_atomic_both
	PASS tlog-test-json-overlay.c:127: two_same_atomics_both
	PASS tlog-test-json-overlay.c:132: object_lower
	PASS tlog-test-json-overlay.c:137: object_upper
	PASS tlog-test-json-overlay.c:142: array_lower
	PASS tlog-test-json-overlay.c:147: array_upper
	PASS tlog-test-json-overlay.c:152: array_same_length
	PASS tlog-test-json-overlay.c:157: array_empty_lower
	PASS tlog-test-json-overlay.c:162: array_empty_upper
	PASS tlog-test-json-overlay.c:167: array_longer_upper
	PASS tlog-test-json-overlay.c:172: array_longer_lower
	PASS tlog-test-json-overlay.c:177: object_in_array_lower
	PASS tlog-test-json-overlay.c:182: object_in_array_upper
	PASS tlog-test-json-overlay.c:187: object_in_array_both
	PASS tlog-test-json-overlay.c:192: nested_atomic_lower
	PASS tlog-test-json-overlay.c:197: nested_atomic_upper
	PASS tlog-test-json-overlay.c:202: nested_same_atomic_both
	PASS tlog-test-json-overlay (exit status: 0)

Do you think that would be better?

Oh, and also please don't make the `line` argument `const`. In general it is nice to do it, but nobody usually does it, so it's confusing to people, and then it makes the code longer.

Thanks!

File: src/tlog-test-fd-json-reader.c [hunk 4]
```diff
@@ -236,7 +236,7 @@ main(void)
          "",
          OP_LOC_GET(1),
          OP_READ(TLOG_RC_OK, NULL),
-         OP_LOC_GET(1),
+         OP_LOC_GET(1),
```
[the remark this answers is not in the document]
Comment (danghai) [reply src/tlog-test-fd-json-reader.c]: Haha the reason why I have this space because I comment `//1` to recovery and create the fail test by putting another value here. I do not realize it has effect on github. Next time, I will open the correct test case on web page

**Comment by danghai** [comment]
Yeah I have same thought with you about issue that we need to make them uniform because it is easy to read. Your idea is very good because FAIL/PASS have the same length. the Function and Line are the same too. It is better to read and keep track of the fail line better when we arrange this format. I will modify it under this format.

**Comment by danghai** [comment]
I modify under new format. I see it is better to read the log file. Please review these change, tell me if you need to change something. The new log fail case:  
[test-suite.log](https://github.com/Scribery/tlog/files/1525067/test-suite.log)

File: src/tlog-test-json-passthrough.c [hunk 5]
```diff
@@ -72,7 +73,8 @@ tlog_test_json_passthrough(const char *name,
              passed;
 
 exit:
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s: %s: %d: %s: \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-json-passthrough.c]: Please don't add `: ` to the end of the line here.

File: src/tlog-test-json-stream-btoa.c [hunk 6]
```diff
@@ -49,7 +49,7 @@ main(void)
                     "%02hhx %02hhx %02hhx %02hhx\n",
                     i,
                     btoa_buf[0], btoa_buf[1],
-                    btoa_buf[2], btoa_buf[3], 
+                    btoa_buf[2], btoa_buf[3],
```
Comment (spbnick) [inline src/tlog-test-json-stream-btoa.c]: Please remove this unnecessary whitespace change.

**Comment by danghai** [comment]
I updated the tlog_debug. When I change the test file, I see I call a lot of `fprintf`. Have you thought about implement debug macro (.h file) and put them under NDEBUG?

File: include/tlog/test_json_stream_enc.h [hunk 7]
```diff
@@ -64,7 +64,7 @@ struct tlog_test_json_stream_enc {
  *
  * @return True if test passed, false otherwise.
  */
-extern bool tlog_test_json_stream_enc(
+extern bool tlog_test_json_stream_enc( const char *file, int line,
```
Comment (spbnick) [inline include/tlog/test_json_stream_enc.h]: Please don't put extra space in front of "const" here.

File: lib/test_json_sink.c [hunk 8]
```diff
@@ -135,7 +135,8 @@ tlog_test_json_sink_run(
 }
 
 bool
-tlog_test_json_sink(const char *name, const struct tlog_test_json_sink test)
+tlog_test_json_sink(const char *file, int line, const char *name, 
+		    const struct tlog_test_json_sink test)
```
Comment (spbnick) [inline lib/test_json_sink.c]: Please align the new line to the inside of the opening parentheses. I.e. put one more space in front of it.

File: lib/test_json_sink.c [hunk 9]
```diff
@@ -159,6 +160,6 @@ tlog_test_json_sink(const char *name, const struct tlog_test_json_sink test)
 
     free(res_output_buf);
 
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n",(passed ? "PASS" : "FAIL"), file, line, name);
```
Comment (spbnick) [inline lib/test_json_sink.c]: Please do not add the space at the end of the format string.

File: lib/test_json_source.c [hunk 10]
```diff
@@ -207,14 +207,15 @@ tlog_test_json_source_run(
...
+tlog_test_json_source(const char *file, int line, const char *name,
                       const struct tlog_test_json_source test)
 {
     bool passed;
 
     passed = tlog_test_json_source_run(name,
                                        test.input, strlen(test.input),
                                        &test.output);
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline lib/test_json_source.c]: Please do not add the space at the end of the format string.

File: lib/test_json_stream_enc.c [hunk 11]
```diff
@@ -157,7 +157,8 @@ tlog_test_json_stream_enc(const char *n,
 #undef CMP_SIZE
 #undef CMP_BOOL
 #undef FAIL
-    fprintf(stderr, "%s: %s\n", n, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline lib/test_json_stream_enc.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-fd-json-reader.c [hunk 12]
```diff
@@ -121,9 +121,10 @@ test(const char *n, const struct test t)
     }
 
 #define FAIL(_fmt, _args...) \
-    do {                                                \
-        fprintf(stderr, "%s: " _fmt "\n", n, ##_args);  \
-        passed = false;                                 \
+    do {                                              \
+        fprintf(stderr, "FAIL %s:%d: %s " _fmt "\n", \
```
Comment (spbnick) [inline src/tlog-test-fd-json-reader.c]: Please align the backslash at the end of this line horizontally with all the others in the macro.

File: src/tlog-test-fd-json-reader.c [hunk 13]
```diff
@@ -192,7 +193,8 @@ test(const char *n, const struct test t)
 #undef FAIL_OP
 #undef FAIL
 
-    fprintf(stderr, "%s: %s\n", n, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d :%s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-fd-json-reader.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-grc.c [hunk 14]
```diff
@@ -31,9 +31,10 @@
 #include <tlog/rc.h>
 
 bool
-test(const char *name, bool res)
+test(const char *file, int line, const char *name, bool res)
 {
-    fprintf(stderr, "%s: %s\n", name, (res ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (res ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-grc.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-json-esc.c [hunk 15]
```diff
@@ -50,9 +51,10 @@ test(const char *name, size_t exp_res_len,
     res_res_len = tlog_json_esc_buf(res_out_buf, out_len, in_ptr, in_len);
 
 #define FAIL(_fmt, _args...) \
-    do {                                                    \
-        fprintf(stderr, "%s: " _fmt "\n", name, ##_args);   \
-        passed = false;                                     \
+    do {                                                            \
+        fprintf(stderr, "FAIL %s:%d: %s " _fmt "\n", file, line,  \
```
Comment (spbnick) [inline src/tlog-test-json-esc.c]: Please align the backslash at the end of this line horizontally with all the others in the macro.

File: src/tlog-test-json-esc.c [hunk 16]
```diff
@@ -69,7 +71,8 @@ test(const char *name, size_t exp_res_len,
     }
 #undef TEST
 #undef FAIL
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"), file,
```
Comment (spbnick) [inline src/tlog-test-json-esc.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-json-esc.c [hunk 17]
```diff
@@ -95,7 +98,7 @@ main(void)
                      _out_ptr, _out_len, _in_str)               \
     TEST_BUF_BUF(_name_token, _exp_res_len,                     \
                  _out_ptr, _out_len, _in_str, strlen(_in_str))
-                 
+
```
Comment (spbnick) [inline src/tlog-test-json-esc.c]: Please do not remove this whitespace in this commit. Make a separate commit, if you want to clean it up.

File: src/tlog-test-json-overlay.c [hunk 18]
```diff
@@ -81,7 +82,8 @@ test(const char *name,
                        exp_result_len);
     }
 
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-json-overlay.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-json-passthrough.c [hunk 19]
```diff
@@ -72,7 +73,8 @@ tlog_test_json_passthrough(const char *name,
              passed;
 
 exit:
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-json-passthrough.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-json-stream-btoa.c (hunk 6 above)
Comment (spbnick) [reply src/tlog-test-json-stream-btoa.c]: This still applies. Please do not remove this extra whitespace in this commit, it's unrelated to the changes you're doing. Make a separate commit cleaning it up (join it with the other unrelated whitespace change above).

File: src/tlog-test-json-stream.c [hunk 20]
```diff
@@ -326,7 +327,8 @@ test(const char *n, const struct test t)
 #undef BUF_CMP
 
 #undef FAIL
-    fprintf(stderr, "%s: %s\n", n, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d: %s \n", (passed ? "PASS" : "FAIL"),
```
Comment (spbnick) [inline src/tlog-test-json-stream.c]: Please do not add the space at the end of the format string.

File: src/tlog-test-json-stream.c [hunk 21]
```diff
@@ -339,7 +341,8 @@ main(void)
     bool passed = true;
 
 #define TEST(_name_token, _struct_init_args...) \
-    passed = test(#_name_token, (struct test){_struct_init_args}) && passed
+    passed = test(__FILE__, __LINE__, #_name_token,     \
+            (struct test){_struct_init_args}) && passed
```
Comment (spbnick) [inline src/tlog-test-json-stream.c]: Please align the beginning of this line to be inside the opening parentheses.

File: src/tlog-test-timespec.c [hunk 22]
```diff
@@ -35,7 +35,7 @@ struct op {
 };
 
 static bool
-op_list_test(const struct op *op_list, size_t op_num,
+op_list_test(const char *file, int line, const struct op *op_list, size_t op_num,
```
Comment (spbnick) [inline src/tlog-test-timespec.c]: Please wrap this line it's getting too long.

File: src/tlog-test-timespec.c [hunk 23]
```diff
@@ -99,9 +101,10 @@ main(void)
 #define TS(_sec, _nsec) (struct timespec){_sec, _nsec}
 
 #define TEST(_op_list, _a, _b, _exp_res) \
-    do {                                                                \
-        passed = op_list_test((_op_list), TLOG_ARRAY_SIZE(_op_list),    \
-                              _a, _b, _exp_res) && passed;              \
+    do {                                                      \
+        passed = op_list_test(__FILE__, __LINE__,(_op_list),  \
```
Comment (spbnick) [inline src/tlog-test-timespec.c]: Please add a space after the second comma here.

**Comment by danghai** [comment]
I modified it based on your comment.

**Comment by danghai** [comment]
I change and review .log file with the format is: STATUS FILE_NAME:LINE_NUMBER TEST_NAME_AND_INFO
I remove the colon after the line number in the output. I squash 2 commits including the unrelated white space. Thanks

File: lib/test_json_sink.c [hunk 24]
```diff
@@ -135,7 +135,8 @@ tlog_test_json_sink_run(
 }
 
 bool
-tlog_test_json_sink(const char *name, const struct tlog_test_json_sink test)
+tlog_test_json_sink(const char *file, int line, const char *name,
+		                const struct tlog_test_json_sink test)
```
Comment (spbnick) [inline lib/test_json_sink.c]: Please align this line to the inside of the opening parentheses. I.e. like this:

    tlog_test_json_sink(const char *file, int line, const char *name,
                        const struct tlog_test_json_sink test)

File: lib/test_json_source.c [hunk 25]
```diff
@@ -207,14 +207,15 @@ tlog_test_json_source_run(
...
                       const struct tlog_test_json_source test)
 {
     bool passed;
 
     passed = tlog_test_json_source_run(name,
                                        test.input, strlen(test.input),
                                        &test.output);
-    fprintf(stderr, "%s: %s\n", name, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d %s\n", (passed ? "PASS" : "FAIL"), 
+	    file, line, name);
```
Comment (spbnick) [inline lib/test_json_source.c]: Please align this line to the inside of opening parentheses.

File: lib/test_json_stream_enc.c [hunk 26]
```diff
@@ -157,7 +157,8 @@ tlog_test_json_stream_enc(const char *n,
 #undef CMP_SIZE
 #undef CMP_BOOL
 #undef FAIL
-    fprintf(stderr, "%s: %s\n", n, (passed ? "PASS" : "FAIL"));
+    fprintf(stderr, "%s %s:%d %s\n", (passed ? "PASS" : "FAIL"), 
+	    file, line, n);
```
Comment (spbnick) [inline lib/test_json_stream_enc.c]: Same here about alignment.

File: src/tlog-test-fd-json-reader.c [hunk 27]
```diff
@@ -217,8 +219,8 @@ main(void)
     {.type = OP_TYPE_LOC_GET,                       \
      .data = {.loc_get = {.exp_loc = _exp_loc}}}
 
-#define TEST(_name_token, _input, _op_list_init_args...) \
-    passed = test(#_name_token,                                 \
+#define TEST(_name_token, _input, _op_list_init_args...)        \
```
Comment (spbnick) [inline src/tlog-test-fd-json-reader.c]: Please leave only one space before the wrapping backslash on the first line of the macro. This makes it easier to see where it starts. Like this:

    #define TEST(_name_token, _input, _op_list_init_args...) \
        passed = test(__FILE__, __LINE__, #_name_token,             \

File: src/tlog-test-json-overlay.c [hunk 28]
```diff
@@ -29,7 +29,8 @@
 #include <assert.h>
 
 bool
-test(const char *name,
+test(const char *file, int line,
```
Comment (spbnick) [inline src/tlog-test-json-overlay.c]: Could you please put `file` and `line` arguments on separate lines to keep to the way other arguments are listed? This way it will be less likely `line` is overlooked.

File: src/tlog-test-json-passthrough.c [hunk 29]
```diff
@@ -41,7 +41,8 @@ struct tlog_test_json_passthrough {
 };
 
 static bool
-tlog_test_json_passthrough(const char *name,
+tlog_test_json_passthrough(const char *file, int line,
```
Comment (spbnick) [inline src/tlog-test-json-passthrough.c]: Same here about arguments on separate lines.

File: src/tlog-test-json-stream.c [hunk 30]
```diff
@@ -339,7 +341,8 @@ main(void)
     bool passed = true;
 
 #define TEST(_name_token, _struct_init_args...) \
-    passed = test(#_name_token, (struct test){_struct_init_args}) && passed
+    passed = test(__FILE__, __LINE__, #_name_token,     \
+                 (struct test){_struct_init_args}) && passed
```
Comment (spbnick) [inline src/tlog-test-json-stream.c]: Same here about aligning to parentheses.

**Comment by spbnick** [comment]
Oh, and please change the commit subject to something clearer. Like "Add file names and line numbers to test output".

**Comment by danghai** [comment]
I realize that the reason makes my editor different look from changes on git is `tab`. If I use `space` instead of `tab`, it does not make any error.

**Comment by spbnick** [comment]
Looks great, merged, thank you, Hai!
